ext/quic/01_quic.js

```diff
@@ -0,0 +1,292 @@
+      yield new QuicBidirectionalStream(r[0], r[1], closed);
+    }
+  } catch (error) {
+    if (ObjectPrototypeIsPrototypeOf(BadResourcePrototype, error)) {
+      return;
+    }
+    throw error;
+  }
+}
+
+async function* uniStream(rid, closed) {
+  try {
+    while (true) {
+      const uniRid = await op_quic_accept_uni(rid);
+      yield readableStream(uniRid, closed);
+    }
+  } catch (error) {
+    if (ObjectPrototypeIsPrototypeOf(BadResourcePrototype, error)) {
+      return;
+    }
+    throw error;
+  }
+}
+
+class QuicConn {
+  #rid = 0;
+  #protocol = null;
+  #bidiStream = null;
+  #uniStream = null;
+  #closed = null;
+
+  constructor(rid, protocol) {
+    this.#rid = rid;
+    this.#protocol = protocol;
+
+    this.#closed = op_quic_connection_closed(this.#rid);
+    core.unrefOpPromise(this.#closed);
+
+    // connection can be indirectly closed by closing listener.
+    SafePromisePrototypeFinally(this.#closed, () => {
+      try {
+        core.close(rid);
+      } catch {
+        // already closed
+      }
+    });
+  }
+
+  get rid() {
+    return this.#rid;
+  }
+
+  get protocol() {
+    return this.#protocol;
+  }
+
+  get remoteAddr() {
+    return op_quic_get_conn_remote_addr(this.#rid);
+  }
+
+  async createBidirectionalStream({ sendOrder } = {}) {
+    const { 0: txRid, 1: rxRid } = await op_quic_open_bi(this.#rid);
+    if (sendOrder != null) {
+      op_quic_set_send_stream_priority(txRid, sendOrder);
+    }
+    return new QuicBidirectionalStream(txRid, rxRid, this.#closed);
+  }
+
+  async createUnidirectionalStream({ sendOrder } = {}) {
+    const rid = await op_quic_open_uni(this.#rid);
+    if (sendOrder != null) {
+      op_quic_set_send_stream_priority(rid, sendOrder);
+    }
+    return writableStream(rid, this.#closed);
+  }
+
+  get incomingBidirectionalStreams() {
+    if (this.#bidiStream == null) {
+      this.#bidiStream = ReadableStream.from(
+        bidiStream(this.#rid, this.#closed),
+      );
+    }
+    return this.#bidiStream;
+  }
+
+  get incomingUnidirectionalStreams() {
+    if (this.#uniStream == null) {
+      this.#uniStream = ReadableStream.from(uniStream(this.#rid, this.#closed));
+    }
+    return this.#uniStream;
+  }
+
+  get maxDatagramSize() {
+    return op_quic_max_datagram_size(this.#rid);
+  }
+
+  async readDatagram(p) {
+    const view = p || new Uint8Array(this.maxDatagramSize);
+    const nread = await op_quic_read_datagram(this.#rid, view);
+    return TypedArrayPrototypeSubarray(view, 0, nread);
+  }
+
+  async sendDatagram(data) {
+    await op_quic_send_datagram(this.#rid, data);
+  }
+
+  get closed() {
+    core.refOpPromise(this.#closed);
+    return this.#closed;
+  }
+
+  close({ closeCode, reason }) {
+    op_quic_close_connection(this.#rid, { closeCode, reason });
+  }
+}
+
+class QuicListener {
+  #rid = 0;
+  #addr = null;
+
+  constructor(rid, addr) {
+    this.#rid = rid;
+    this.#addr = addr;
+  }
+
+  get rid() {
+    return this.#rid;
+  }
```

Comment:
Do you need to access `rid` somewhere in code? We're moving away (as in deprecating and removing) APIs that accept `rid` as a parameter (we want to keep it an internal notion).

Reply:
was just copying from other stuff, will remove :)